signer/threshold_signer_soft.go

```diff
@@ -0,0 +1,347 @@
+	meta.Peers[softSigner.Key.ID-1].EphemeralSecretPublicKey = ourEphPublicKey
+
+	// grab the peer info for the ID being requested
+	peer, ok := peers[req.ID]
+	if !ok {
+		return res, errors.New("unknown peer ID")
+	}
+
+	sharePart := meta.DealtShares[req.ID-1]
+
+	// use RSA public to encrypt user's share part
+	encrypted, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, &peer.PublicKey, sharePart, nil)
+	if err != nil {
+		return res, err
+	}
+
+	res.SourceID = softSigner.Key.ID
+	res.SourceEphemeralSecretPublicKey = ourEphPublicKey
+	res.EncryptedSharePart = encrypted
+
+	// sign the response payload with our private key
+	// cosigners can verify the signature to confirm sender validity
+
+	jsonBytes, err := tmjson.Marshal(res)
+
+	if err != nil {
+		return res, err
+	}
+
+	digest := sha256.Sum256(jsonBytes)
+	signature, err := rsa.SignPSS(rand.Reader, &softSigner.Key.RSAKey, crypto.SHA256, digest[:], nil)
+
+	if err != nil {
+		return res, err
+	}
+
+	res.SourceSig = signature
+
+	res.DestinationID = req.ID
+
+	return res, nil
+}
+
+// Store an ephemeral secret share part provided by another cosigner (signer)
+// Implements ThresholdSigner
+func (softSigner *ThresholdSignerSoft) SetEphemeralSecretPart(
+	req CosignerSetEphemeralSecretPartRequest, m *LastSignStateStruct, peers map[int]CosignerPeer) error {
+
+	// Verify the source signature
+	if req.SourceSig == nil {
+		return errors.New("SourceSig field is required")
+	}
+
+	digestMsg := CosignerEphemeralSecretPart{}
+	digestMsg.SourceID = req.SourceID
+	digestMsg.SourceEphemeralSecretPublicKey = req.SourceEphemeralSecretPublicKey
+	digestMsg.EncryptedSharePart = req.EncryptedSharePart
+
+	digestBytes, err := tmjson.Marshal(digestMsg)
+	if err != nil {
+		return err
+	}
+
+	digest := sha256.Sum256(digestBytes)
+	peer, ok := peers[req.SourceID]
+
+	if !ok {
+		return fmt.Errorf("unknown cosigner: %d", req.SourceID)
+	}
+
+	peerPub := peer.PublicKey
+	err = rsa.VerifyPSS(&peerPub, crypto.SHA256, digest[:], req.SourceSig, nil)
+	if err != nil {
+		return err
+	}
+
+	// protects the meta map
+	m.LastSignStateMutex.Lock()
+	defer m.LastSignStateMutex.Unlock()
+
+	hrst := HRSTKey{
+		Height:    req.Height,
+		Round:     req.Round,
+		Step:      req.Step,
+		Timestamp: req.Timestamp.UnixNano(),
+	}
+
+	meta, ok := softSigner.HrsMeta[hrst]
+	// generate metadata placeholder
+	if !ok {
+		newMeta, err := softSigner.DealShares(CosignerGetEphemeralSecretPartRequest{
+			Height: req.Height,
+			Round:  req.Round,
+			Step:   req.Step,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		meta = newMeta
+		softSigner.HrsMeta[hrst] = meta
+	}
+
+	// decrypt share
+	sharePart, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, &softSigner.Key.RSAKey, req.EncryptedSharePart, nil)
+	if err != nil {
+		return err
+	}
+
+	// set slot
+	meta.Peers[req.SourceID-1].Share = sharePart
+	meta.Peers[req.SourceID-1].EphemeralSecretPublicKey = req.SourceEphemeralSecretPublicKey
```

Comment:
Ensuring that these statements are going to have an effect, is very subtle.

The HrsMeta map has value types (not pointer types). 

After assigning meta into the map on L334, we are then modifying the local value. Luckily we are indexing into an existing slice, so our value and the value in the map should reference the same slice. Then, we are overwriting the two fields of the value at that index in the slice. Does this write the underlying value in the shared struct? I think so, but it requires careful reading to follow.

It would be much better to have a test demonstrating that this works properly, than requiring the user to follow the code.

Reply:
@mark-rushakoff (and @agouin).
I am having trouble with finding the best way to fix this! 
I 100% agree with @mark-rushakoff  its very subtile and makes the code  almost"impossible" to read and not even a seasoned gopher as you Mark are able to get along. However I am not sure what you mean with "test demonstrating it works". Pure test or a "type casting fix" or a check. 

My solution to fix this is either to use some kind of setter function/method or rewrite and check for softSigner.HrsMeta[hrst].Peers[req.SourceID-1].Share == sharePart or another way to ensure typecasting is maybe to do copy(meta.Peers[req.SourceID-1].Share, sharePart)

Let me know what you think. Its very very very subtile that this is actually legit.